src/main/flight/position.c

```diff
@@ -22,205 +22,202 @@
-            }
+        if (isGpsAvailable && positionConfig()->altitude_source != ALTITUDE_SOURCE_BARO_ONLY) {
+            gpsAltGroundCm = gpsAltCm;      // watch for valid GPS altitude data to zero GPS altitude with
+            displayAltitudeCm = gpsAltCm;   // estimatedAltitude shows most recent ASL GPS altitude in OSD and sensors, while disarmed
+        } else if (isBaroAvailable && positionConfig()->altitude_source != ALTITUDE_SOURCE_GPS_ONLY) {
+            displayAltitudeCm = baroAltCm;
         }
-        zeroedAltitudeCm = 0.0f; // always hold relativeAltitude at zero while disarmed
-        DEBUG_SET(DEBUG_ALTITUDE, 2, gpsAltCm / 100.0f); // Absolute altitude ASL in metres, max 32,767m
+
+        altitudeCm = 0.0f; // always hold relative altitude at zero while disarmed
+
     //  ***  ARMED  ***
     } else {
-        if (!wasArmed) { // things to run once, on arming, after being disarmed
-            baroAltOffsetCm = newBaroAltOffsetCm;
+
+        // things to run once, on arming, after being disarmed
+        if (!wasArmed) {
+#ifdef USE_BARO
+            baroSetGroundLevel();  // zero barometer altitude
+#endif
             wasArmed = true;
         }
 
-        baroAltCm -= baroAltOffsetCm; // use smoothed baro with most recent zero from disarm period
-
-        if (haveGpsAlt) { // update relativeAltitude with every new gpsAlt value, or hold the previous value until 3D lock recovers
-            if (!useZeroedGpsAltitude && haveBaroAlt) { // armed without zero offset, can use baro values to zero later
-                gpsAltOffsetCm = gpsAltCm - baroAltCm; // not very accurate
-                useZeroedGpsAltitude = true;
-            }
-            if (useZeroedGpsAltitude) { // normal situation
-                zeroedAltitudeCm = gpsAltCm - gpsAltOffsetCm; // now that we have a GPS offset value, we can use it to zero relativeAltitude
-            }
-        } else {
-            gpsTrust = 0.0f;
-            // TO DO - smoothly reduce GPS trust, rather than immediately dropping to zero for what could be only a very brief loss of 3D fix 
-        }
-        DEBUG_SET(DEBUG_ALTITUDE, 2, lrintf(zeroedAltitudeCm / 10.0f)); // Relative altitude above takeoff, to 0.1m, rolls over at 3,276.7m
-        
-        // Empirical mixing of GPS and Baro altitudes
-        if (useZeroedGpsAltitude && (positionConfig()->altitude_source == DEFAULT || positionConfig()->altitude_source == GPS_ONLY)) {
-            if (haveBaroAlt && positionConfig()->altitude_source == DEFAULT) {
-                // mix zeroed GPS with Baro altitude data, if Baro data exists if are in default altitude control mode 
-                const float absDifferenceM = fabsf(zeroedAltitudeCm - baroAltCm) / 100.0f * positionConfig()->altitude_prefer_baro / 100.0f;
-                if (absDifferenceM > 1.0f) { // when there is a large difference, favour Baro
-                    gpsTrust /=  absDifferenceM;
+        switch (positionConfig()->altitude_source) {
```

Comment:
Maybe:
select actually used source first (first switch or if chain), then use it to select value.
It will remove code duplication ( altitudeCm = <source> ) and will prevent using gps/baro when not available. And it will remove inconsistency between ALTITUDE_SOURCE_DEFAULT and ALTITUDE_SOURCE_*_ONLY

Reply:
The code is written this way to make sure we have a fallback option dynamically in flight, even with `ALTITUDE_SOURCE_DEFAULT` selected (static CLI setting).

Maybe I don't understand your idea but I made it like this on purpose, including the code duplication. But feel free to make a pseudo code suggestion.

Reply:
```
source = NONE;
switch (positionConfig()->altitude_source) {
  case ALTITUDE_SOURCE_DEFAULT:
     if (isGpsAvailable && isBaroAvailable) {
             source = FUSE;
     } else if (isGpsAvailable) {
        source = GPS;
     } else if (isBaroAvailable) {
        source = BARO;
     } else {
        source = NONE;
     }
     break;
case ALTITUDE_SOURCE_GPS_ONLY:
   source = isGpsAvailable ? GPS : NONE:
   break;
case ALTITUDE_SOURCE_BARO_ONLY
   source = isBaroAvailable ? BARO : NONE:
   break;
}

switch(source) {
```

You can easily add modes like 'use GPS if available, else baro (no fuse)